compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -75,186 +75,490 @@ std::optional<uint32_t> getNumberIterations(scf::ForOp loop) {
-            getGeneratorTileOp<CopyOpOperateOn::Source>(npuDmaOp,
-                                                        shimTileToGeneratorMap);
-        if (succeeded(tile) && *tile == tileOp) bdIdCount++;
-      }
-      if (npuDmaOp.getTarget()) {
-        FailureOr<AMDAIE::TileOp> tile =
-            getGeneratorTileOp<CopyOpOperateOn::Target>(npuDmaOp,
-                                                        shimTileToGeneratorMap);
-        if (succeeded(tile) && *tile == tileOp) bdIdCount++;
+/// A DmaBatch contains the following :-
+///   1. DmaOps within the current batch.
+///   2. Required Bd Ids by the current batch.
+///   3. Nested DmaBatch chain.
+///   4. Pointer to the sibling DmaBatch.
+///   4. A pointer to the parent op if the parent of is a scf.for.
+struct DmaBatch {
+  SmallVector<AMDAIE::NpuDmaCpyNdOp> currentDmaOps = {};
+  int32_t requiredBdIds = 0;
+  SmallVector<std::shared_ptr<DmaBatch>> immediateInnerBatches = {};
+  std::shared_ptr<DmaBatch> nextDmaBatch = nullptr;
+  scf::ForOp forOpParent = nullptr;
+
+  /// A DmaBatch with no DmaOps and no nested DmaBatches is an empty DmaBatch.
+  bool isEmpty() const {
+    return (currentDmaOps.empty() && immediateInnerBatches.empty());
+  }
+
+  /// Given a DmaBatch `outerDmaBatch`, traverse each nested DmaBatch in it and
+  /// infer the required Bd Ids for them. If the nested DmaBatch is surrounded
+  /// with a scf.for, we account for that while sending the required Bd Ids to
+  /// the caller of this API.
+  int32_t getRequiredBdIdsForInnerBatches() const {
+    if (immediateInnerBatches.empty()) return 0;
+    int32_t requiredBdIds = 0;
+    for (const std::shared_ptr<DmaBatch> &dmaBatch : immediateInnerBatches) {
+      std::shared_ptr<DmaBatch> currDmaBatch = dmaBatch;
+      do {
+        if (currDmaBatch->isEmpty()) break;
+        int32_t totalDmaOpsInCurrentBatch = currDmaBatch->currentDmaOps.size();
+        int32_t requiredBdIdsForInnerBatches =
+            currDmaBatch->getRequiredBdIdsForInnerBatches();
+        currDmaBatch->requiredBdIds =
+            requiredBdIdsForInnerBatches + totalDmaOpsInCurrentBatch;
+        requiredBdIds += totalDmaOpsInCurrentBatch;
+        if (currDmaBatch->forOpParent) {
+          std::optional<uint32_t> numIterations =
+              getNumberIterations(currDmaBatch->forOpParent);
+          if (numIterations) {
+            requiredBdIds +=
+                requiredBdIdsForInnerBatches * numIterations.value();
+          } else {
+            requiredBdIds += requiredBdIdsForInnerBatches;
+          }
+        }
+        currDmaBatch = currDmaBatch->nextDmaBatch;
+      } while (currDmaBatch != nullptr);
+    }
+    return requiredBdIds;
+  }
+};
+
+/// `TileDmaBatchGraph` is a struct that maintains a mapping from a tile to its
+/// corresponding `DmaBatch`. It contains utilities that help form/update the
+/// `DmaBatch` graph for a particular tile.
+struct TileDmaBatchGraph {
+ private:
```

Comment:
```suggestion
class TileDmaBatchGraph {
```
I think a class is more appropriate as you're making use of private member variables. Then you can remove `private:` here as well.